### ISSUE TITLE
[BUG] MPICH in conda binary release incompatible with mpi4py and horovod

### ISSUE THREAD

**Issue by chazeon** [deepmodeling/deepmd-kit#1090]
**Summary**

In the prebuilt conda environment in the release, LAMMPS and PLUMMED are depended on MPICH.

However, horovod as is used for parallel training does not support MPICH in general and supports OpenMPI only, see [this issue](https://github.com/horovod/horovod/issues/1637).

The parallel environments will not be correctly detected when using external OpenMPI, see [this](https://stackoverflow.com/questions/29264640/mpiexec-and-python-mpi4py-gives-rank-0-and-size-1) and [this](https://gist.github.com/ozooxo/9280304).

This cause all process run on the same GPU and run out of GPU memory and cause `CUDA_ERROR_OUT_OF_MEMORY` as described on the [horovod's troubleshooting page](https://horovod.readthedocs.io/en/stable/troubleshooting_include.html#running-out-of-memory).

Removing MPICH means I will also need to remove the prebuilt LAMMPS and PLUMMED.

The prebuilt package might needs to be updated with an OpenMPI version.

To avoid this problem, I am planning to try horovod with Gloo, not sure if it works or not.

**Deepmd-kit version, installation way, input file, running commands, error log, etc.**

- Deepmd-kit version: 2.0.0
- Installation: from release binary

Additional conda packages:

```bash
module purge
module load gpu

eval "$(/home/chazeon/SCRATCH/deepmd-kit-2.0.0-cuda10.1_gpu/bin/conda shell.bash hook)"

conda install -c anaconda gxx_linux-64
conda install -c anaconda cmake
pip install mpi4py
HOROVOD_WITHOUT_GLOO=1 HOROVOD_WITH_TENSORFLOW=1 pip install horovod
```

**Steps to Reproduce**

Code for a quick test is the following (`test.py`):

```python
import mpi4py
from mpi4py import MPI

comm = MPI.COMM_WORLD
size = comm.Get_size()
rank = comm.Get_rank()

print("Rank:", rank)
print("Size:", size)

print(mpi4py.get_config())
```

```
(base) [chazeon@exp-7-59 000.1]$ python3 test.py
Rank: 0
Size: 1
{'mpicc': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpicc', 'mpicxx': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpic++', 'mpif90': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif90', 'mpif77': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif77'}
(base) [chazeon@exp-7-59 000.1]$ mpirun -n 2 python3 test.py
Rank: 0
Size: 1
Rank: 0
Size: 1
{'mpicc': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpicc', 'mpicxx': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpic++', 'mpif90': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif90', 'mpif77': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif77'}
{'mpicc': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpicc', 'mpicxx': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpic++', 'mpif90': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif90', 'mpif77': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif77'}
```

Because of this, the following code

```python
import horovod.tensorflow as HVD

HVD.init()
print(HVD.size())

```
always gives `1`.

This will eventually affect `_try_init_distrib` and `is_distrib` because `HVD.size()` always gives `1`, in `deepmd/train/run_options.py` line 177, `self.is_distrib` is always `False`.

```python
    def _try_init_distrib(self):
        try:
            import horovod.tensorflow as HVD
            HVD.init()
            self.is_distrib = HVD.size() > 1
        except ImportError:
            log.warning("Switch to serial execution due to lack of horovod module.")
            self.is_distrib = False

        # Do real intialization
        if self.is_distrib:
            self._init_distributed(HVD)
            self._HVD = HVD
        else:
            self._init_serial()
            self._HVD = None
```

This is likely because of MPICH, see [this](https://stackoverflow.com/questions/29264640/mpiexec-and-python-mpi4py-gives-rank-0-and-size-1) and [this](https://gist.github.com/ozooxo/9280304).

This eventually causes multiple processes are running without knowing each other, they try to exhaust the GPU memory on the same GPU card similar the description on the [horovod's troubleshooting page](https://horovod.readthedocs.io/en/stable/troubleshooting_include.html#running-out-of-memory).

See discussion #1089 for full log.

```
(base) [chazeon@exp-7-59 000.1]$ CUDA_VISIBLE_DEVICES=0,1 horovodrun -np 2 dp train params.yml
...
[1,0]<stderr>:DEEPMD INFO    training data with min nbor dist: 1.0231442946289506
[1,0]<stderr>:DEEPMD INFO    training data with max nbor size: [64, 134, 64]
[1,1]<stderr>:DEEPMD INFO    training data with min nbor dist: 1.0231442946289506
[1,1]<stderr>:DEEPMD INFO    training data with max nbor size: [64, 134, 64]
[1,1]<stderr>:DEEPMD INFO     _____               _____   __  __  _____           _     _  _
[1,1]<stderr>:DEEPMD INFO    |  __ \             |  __ \ |  \/  ||  __ \         | |   (_)| |
[1,1]<stderr>:DEEPMD INFO    | |  | |  ___   ___ | |__) || \  / || |  | | ______ | | __ _ | |_
[1,1]<stderr>:DEEPMD INFO    | |  | | / _ \ / _ \|  ___/ | |\/| || |  | ||______|| |/ /| || __|
[1,1]<stderr>:DEEPMD INFO    | |__| ||  __/|  __/| |     | |  | || |__| |        |   < | || |_
[1,1]<stderr>:DEEPMD INFO    |_____/  \___| \___||_|     |_|  |_||_____/         |_|\_\|_| \__|
[1,1]<stderr>:DEEPMD INFO    Please read and cite:
[1,1]<stderr>:DEEPMD INFO    Wang, Zhang, Han and E, Comput.Phys.Comm. 228, 178-184 (2018)
[1,1]<stderr>:DEEPMD INFO    installed to:         /tmp/pip-req-build-m0vk2rxi/_skbuild/linux-x86_64-3.9/cmake-install
[1,1]<stderr>:DEEPMD INFO    source :              v2.0.0
[1,1]<stderr>:DEEPMD INFO    source brach:         HEAD
[1,1]<stderr>:DEEPMD INFO    source commit:        1a25414
[1,1]<stderr>:DEEPMD INFO    source commit at:     2021-08-28 08:15:38 +0800
[1,1]<stderr>:DEEPMD INFO    build float prec:     double
[1,1]<stderr>:DEEPMD INFO    build with tf inc:    /expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/lib/python3.9/site-packages/tensorflow/include
[1,1]<stderr>:DEEPMD INFO    build with tf lib:
[1,1]<stderr>:DEEPMD INFO    ---Summary of the training---------------------------------------
[1,1]<stderr>:DEEPMD INFO    running on:           exp-7-59
[1,1]<stderr>:DEEPMD INFO    computing device:     gpu:0
[1,1]<stderr>:DEEPMD INFO    CUDA_VISIBLE_DEVICES: 0,1
[1,1]<stderr>:DEEPMD INFO    Count of visible GPU: 2
[1,1]<stderr>:DEEPMD INFO    num_intra_threads:    0
[1,1]<stderr>:DEEPMD INFO    num_inter_threads:    0
[1,1]<stderr>:DEEPMD INFO    -----------------------------------------------------------------
[1,0]<stderr>:DEEPMD INFO     _____               _____   __  __  _____           _     _  _
[1,0]<stderr>:DEEPMD INFO    |  __ \             |  __ \ |  \/  ||  __ \         | |   (_)| |
[1,0]<stderr>:DEEPMD INFO    | |  | |  ___   ___ | |__) || \  / || |  | | ______ | | __ _ | |_
[1,0]<stderr>:DEEPMD INFO    | |  | | / _ \ / _ \|  ___/ | |\/| || |  | ||______|| |/ /| || __|
[1,0]<stderr>:DEEPMD INFO    | |__| ||  __/|  __/| |     | |  | || |__| |        |   < | || |_
[1,0]<stderr>:DEEPMD INFO    |_____/  \___| \___||_|     |_|  |_||_____/         |_|\_\|_| \__|
[1,0]<stderr>:DEEPMD INFO    Please read and cite:
[1,0]<stderr>:DEEPMD INFO    Wang, Zhang, Han and E, Comput.Phys.Comm. 228, 178-184 (2018)
[1,0]<stderr>:DEEPMD INFO    installed to:         /tmp/pip-req-build-m0vk2rxi/_skbuild/linux-x86_64-3.9/cmake-install
[1,0]<stderr>:DEEPMD INFO    source :              v2.0.0
[1,0]<stderr>:DEEPMD INFO    source brach:         HEAD
[1,0]<stderr>:DEEPMD INFO    source commit:        1a25414
[1,0]<stderr>:DEEPMD INFO    source commit at:     2021-08-28 08:15:38 +0800
[1,0]<stderr>:DEEPMD INFO    build float prec:     double
[1,0]<stderr>:DEEPMD INFO    build with tf inc:    /expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/lib/python3.9/site-packages/tensorflow/include
[1,0]<stderr>:DEEPMD INFO    build with tf lib:
[1,0]<stderr>:DEEPMD INFO    ---Summary of the training---------------------------------------
[1,0]<stderr>:DEEPMD INFO    running on:           exp-7-59
[1,0]<stderr>:DEEPMD INFO    computing device:     gpu:0
[1,0]<stderr>:DEEPMD INFO    CUDA_VISIBLE_DEVICES: 0,1
[1,0]<stderr>:DEEPMD INFO    Count of visible GPU: 2
[1,0]<stderr>:DEEPMD INFO    num_intra_threads:    0
[1,0]<stderr>:DEEPMD INFO    num_inter_threads:    0
[1,0]<stderr>:DEEPMD INFO    -----------------------------------------------------------------

...

[1,0]<stderr>:2021-09-02 12:44:54.392611: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 29.84G (32039239680 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.393996: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 26.85G (28835315712 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.395343: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 24.17G (25951782912 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.396857: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 21.75G (23356604416 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.398347: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 19.58G (21020944384 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.399878: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 17.62G (18918848512 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.401414: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 15.86G (17026963456 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.402936: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 14.27G (15324266496 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.406219: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 12.84G (13791839232 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.409880: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 11.56G (12412654592 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.412509: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 10.40G (11171388416 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.416159: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 9.36G (10054249472 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.418764: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 8.43G (9048823808 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.422393: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 7.58G (8143941120 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.424980: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 6.83G (7329546752 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.428616: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 6.14G (6596592128 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.431204: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 5.53G (5936932864 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.433794: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 4.98G (5343239680 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.437447: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 4.48G (4808915456 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.440039: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 4.03G (4328023552 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.443686: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 3.63G (3895220992 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.446268: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 3.26G (3505698816 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.449884: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 2.94G (3155128832 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.452390: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 2.64G (2839616000 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.454955: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 2.38G (2555654400 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.458543: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 2.14G (2300088832 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.461126: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 1.93G (2070080000 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.464731: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 1.73G (1863072000 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.467280: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 1.56G (1676764928 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.470896: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 1.41G (1509088512 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.473391: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 1.26G (1358179584 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.475956: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 1.14G (1222361600 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
[1,0]<stderr>:2021-09-02 12:44:54.479581: I tensorflow/stream_executor/cuda/cuda_driver.cc:789] failed to allocate 1.02G (1100125440 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY: out of memory
```

Removing MPICH means I will also need to remove the prebuilt LAMMPS and PLUMMED, it gives the following:

```
-> % conda remove mpich
Collecting package metadata (repodata.json): done
Solving environment: |
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - deepmodeling/linux-64::deepmd-kit==2.0.0=py39_0_cuda10.1_gpu
done

## Package Plan ##

  environment location: /expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu

  removed specs:
    - mpich


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    binutils_impl_linux-64-2.33.1|       he6710b0_7         3.7 MB
    binutils_linux-64-2.33.1   |      h9595d00_15          26 KB
    bzip2-1.0.8                |       h7b6447c_0          78 KB
    ca-certificates-2021.7.5   |       h06a4308_1         113 KB
    expat-2.4.1                |       h2531618_2         168 KB
    gcc_impl_linux-64-7.3.0    |       habb00fd_1        41.9 MB
    gcc_linux-64-7.3.0         |      h553295d_15          27 KB
    gxx_impl_linux-64-7.3.0    |       hdf63c60_1        15.0 MB
    gxx_linux-64-7.3.0         |      h553295d_15          26 KB
    krb5-1.19.2                |       hac12032_0         1.2 MB
    ld_impl_linux-64-2.33.1    |       h53a641e_7         568 KB
    libcurl-7.78.0             |       h0b77cf5_0         338 KB
    libedit-3.1.20210714       |       h7f8727e_0         165 KB
    libev-4.33                 |       h7b6447c_0         112 KB
    libnghttp2-1.41.0          |       hf8bcb03_2         667 KB
    libssh2-1.9.0              |       h1ba5d50_1         269 KB
    libuv-1.40.0               |       h7b6447c_0         736 KB
    lz4-c-1.9.3                |       h295c915_1         185 KB
    openssl-1.1.1l             |       h7f8727e_0         2.5 MB
    pyyaml-5.4.1               |   py39h27cfd23_1         174 KB
    rhash-1.4.1                |       h3c74f83_1         203 KB
    zstd-1.4.5                 |       h9ceee32_0         619 KB
    ------------------------------------------------------------
                                           Total:        68.6 MB

The following NEW packages will be INSTALLED:

  libev              pkgs/main/linux-64::libev-4.33-h7b6447c_0
  libnghttp2         pkgs/main/linux-64::libnghttp2-1.41.0-hf8bcb03_2
  pyyaml             pkgs/main/linux-64::pyyaml-5.4.1-py39h27cfd23_1

The following packages will be REMOVED:

  lammps-dp-2.0.0-h9954876_0
  mpich-3.3.2-hc856adb_0
  plumed-2.6.2-h5c0ae08_0

The following packages will be UPDATED:

  ca-certificates    anaconda::ca-certificates-2020.10.14-0 --> pkgs/main::ca-certificates-2021.7.5-h06a4308_1
  expat                   anaconda::expat-2.2.10-he6710b0_2 --> pkgs/main::expat-2.4.1-h2531618_2
  krb5                     anaconda::krb5-1.18.2-h173b8e3_0 --> pkgs/main::krb5-1.19.2-hac12032_0
  libcurl               anaconda::libcurl-7.71.1-h20c2e04_1 --> pkgs/main::libcurl-7.78.0-h0b77cf5_0
  libedit            anaconda::libedit-3.1.20191231-h14c39~ --> pkgs/main::libedit-3.1.20210714-h7f8727e_0
  lz4-c                    anaconda::lz4-c-1.9.2-heb0550a_3 --> pkgs/main::lz4-c-1.9.3-h295c915_1
  openssl                                 1.1.1k-h27cfd23_0 --> 1.1.1l-h7f8727e_0
  rhash                    anaconda::rhash-1.4.0-h1ba5d50_0 --> pkgs/main::rhash-1.4.1-h3c74f83_1

The following packages will be SUPERSEDED by a higher-priority channel:

  bzip2                                            anaconda --> pkgs/main
  libssh2                                          anaconda --> pkgs/main
  libuv                                            anaconda --> pkgs/main
  zstd                                             anaconda --> pkgs/main

The following packages will be DOWNGRADED:

  binutils_impl_lin~                      2.35.1-h27ae35d_9 --> 2.33.1-he6710b0_7
  binutils_linux-64                      2.35.1-h454624a_30 --> 2.33.1-h9595d00_15
  gcc_impl_linux-64                       9.3.0-h6df7d76_17 --> 7.3.0-habb00fd_1
  gcc_linux-64                            9.3.0-h1ee779e_30 --> 7.3.0-h553295d_15
  gxx_impl_linux-64                       9.3.0-hbdd7822_17 --> 7.3.0-hdf63c60_1
  gxx_linux-64                            9.3.0-h7e70986_30 --> 7.3.0-h553295d_15
  ld_impl_linux-64                        2.35.1-h7274673_9 --> 2.33.1-h53a641e_7


Proceed ([y]/n)?
```

**Further Information, Files, and Links**

**Comment by amcadmus** [comment]
@shishaochen @njzjz Could you plz take a look?

**Comment by shishaochen** [comment]
> @shishaochen @njzjz Could you plz take a look?

Sure. I'll reproduce first.

**Comment by njzjz** [comment]
Your mpi4py is built against openmpi:

```
{'mpicc': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpicc', 'mpicxx': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpic++', 'mpif90': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif90', 'mpif77': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif77'}
```

This, of course, would work wrongly if you mixed it with mpich. Per [this page](https://mpi4py.readthedocs.io/en/1.3.1/install.html), you need to specify the correct `MPICC`.

You may consider using conda-forge's prebuilt mpi4py

```sh
conda install mpi4py -c defaults -c conda-forge
```

where `test.py` will work fine.
```
$ python test.py
Rank: 0
Size: 1
{'mpicc': '/home/jz748/deepmd-kit/bin/mpicc', 'mpicxx': '/home/jz748/deepmd-kit/bin/mpicxx', 'mpifort': '/home/jz748/deepmd-kit/bin/mpifort', 'mpif90': '/home/jz748/deepmd-kit/bin/mpif90', 'mpif77': '/home/jz748/deepmd-kit/bin/mpif77'}
$ mpirun -n 2 python test.py
Rank: 0
Size: 2
{'mpicc': '/home/jz748/deepmd-kit/bin/mpicc', 'mpicxx': '/home/jz748/deepmd-kit/bin/mpicxx', 'mpifort': '/home/jz748/deepmd-kit/bin/mpifort', 'mpif90': '/home/jz748/deepmd-kit/bin/mpif90', 'mpif77': '/home/jz748/deepmd-kit/bin/mpif77'}
Rank: 1
Size: 2
{'mpicc': '/home/jz748/deepmd-kit/bin/mpicc', 'mpicxx': '/home/jz748/deepmd-kit/bin/mpicxx', 'mpifort': '/home/jz748/deepmd-kit/bin/mpifort', 'mpif90': '/home/jz748/deepmd-kit/bin/mpif90', 'mpif77': '/home/jz748/deepmd-kit/bin/mpif77'}
```

**Comment by chazeon** [comment]
Hi @njzjz,

> Your mpi4py is built against openmpi:
> 
> ```
> {'mpicc': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpicc', 'mpicxx': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpic++', 'mpif90': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif90', 'mpif77': '/cm/shared/apps/spack/gpu/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/openmpi-4.0.4-mpm64p2fdhmmvnb5v2gszma2fsgdyu73/bin/mpif77'}
> ```
> 
> This, of course, would work wrongly if you mixed it with mpich. Per [this page](https://mpi4py.readthedocs.io/en/1.3.1/install.html), you need to specify the correct `MPICC`.
> 
> You may consider using conda-forge's prebuilt mpi4py
> 
> ```shell
> conda install mpi4py -c defaults -c conda-forge
> ```
> 
> where `test.py` will work fine.
> 
> ```
> $ python test.py
> Rank: 0
> Size: 1
> {'mpicc': '/home/jz748/deepmd-kit/bin/mpicc', 'mpicxx': '/home/jz748/deepmd-kit/bin/mpicxx', 'mpifort': '/home/jz748/deepmd-kit/bin/mpifort', 'mpif90': '/home/jz748/deepmd-kit/bin/mpif90', 'mpif77': '/home/jz748/deepmd-kit/bin/mpif77'}
> $ mpirun -n 2 python test.py
> Rank: 0
> Size: 2
> {'mpicc': '/home/jz748/deepmd-kit/bin/mpicc', 'mpicxx': '/home/jz748/deepmd-kit/bin/mpicxx', 'mpifort': '/home/jz748/deepmd-kit/bin/mpifort', 'mpif90': '/home/jz748/deepmd-kit/bin/mpif90', 'mpif77': '/home/jz748/deepmd-kit/bin/mpif77'}
> Rank: 1
> Size: 2
> {'mpicc': '/home/jz748/deepmd-kit/bin/mpicc', 'mpicxx': '/home/jz748/deepmd-kit/bin/mpicxx', 'mpifort': '/home/jz748/deepmd-kit/bin/mpifort', 'mpif90': '/home/jz748/deepmd-kit/bin/mpif90', 'mpif77': '/home/jz748/deepmd-kit/bin/mpif77'}
> ```

I can reproduce this behavior and I can confirm that `mpi4py` works with MPICH on a login node.

```
(base) chazeon@x86_64-conda-linux-gnu [08:56:54 AM] [~/SCRATCH/qe-jobs/20210823-AlOOH-dp_init/dp_run/iter.000000/00.train/000.1]
-> % mpirun -np 2 python test.py
Rank: 0
Size: 2
{'mpicc': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpicc', 'mpicxx': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpicxx', 'mpifort': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpifort', 'mpif90': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpif90', 'mpif77': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpif77'}
Rank: 1
Size: 2
{'mpicc': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpicc', 'mpicxx': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpicxx', 'mpifort': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpifort', 'mpif90': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpif90', 'mpif77': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpif77'}
(base) chazeon@x86_64-conda-linux-gnu [08:57:03 AM] [~/SCRATCH/qe-jobs/20210823-AlOOH-dp_init/dp_run/iter.000000/00.train/000.1]
-> % python test.py
Rank: 0
Size: 1
{'mpicc': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpicc', 'mpicxx': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpicxx', 'mpifort': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpifort', 'mpif90': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpif90', 'mpif77': '/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/bin/mpif77'}
```
I am actually having some problem with using conda's MPICH on a GPU node
```
(base) [chazeon@exp-7-60 000.1]$ SLURM_TASKS_PER_NODE=2 mpirun -np 1  python test.py -v
Error: node list format not recognized. Try using '-hosts=<hostnames>'.
Aborted (core dumped)
```

Nevertheless, the origin issue is then horovod does not work with MPICH
```
(base) [chazeon@exp-7-60 000.1]$ cat test_hvd.py
import horovod.tensorflow as HVD
HVD.init()
print("HVD size:", HVD.size())
(base) [chazeon@exp-7-60 000.1]$ python test_hvd.py
2021-09-03 09:32:41.535033: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
HVD size: 1
(base) [chazeon@exp-7-60 000.1]$ horovodrun -np 2 python test_hvd.py
2021-09-03 09:31:41.992620: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
[mpiexec@exp-7-60] match_arg (utils/args/args.c:163): unrecognized argument allow-run-as-root
[mpiexec@exp-7-60] HYDU_parse_array (utils/args/args.c:178): argument matching returned error
[mpiexec@exp-7-60] parse_args (ui/mpich/utils.c:1642): error parsing input array
[mpiexec@exp-7-60] HYD_uii_mpx_get_parameters (ui/mpich/utils.c:1694): unable to parse user arguments
[mpiexec@exp-7-60] main (ui/mpich/mpiexec.c:148): error parsing parameters
```

Like @shishaochen said, I might need to use a separate conda environment without MPICH for the training, there is no prebuilt one currently, so I need to manually create one I guess

**Comment by njzjz** [comment]
Fixed by horovod/horovod#3148.

**Comment by chazeon** [comment]
Thanks, I will check out this horovod fix tomorrow.

Meanwhile, I am trying to build horovod with Gloo.

```bash
HOROVOD_WITH_GLOO=1 \
HOROVOD_WITH_TENSORFLOW=1 \
HOROVOD_WITHOUT_MPI=1 \
HOROVOD_GPU_OPERATIONS=NCCL \
pip install horovod \
	--force-reinstall --ignore-installed \
	--no-binary horovod \
	--verbose
```

Then we have
```
(base) [chazeon@exp-7-59 000.1]$ horovodrun --check-build
2021-09-03 20:21:40.394236: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
2021-09-03 20:21:57.904508: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
2021-09-03 20:22:01.341202: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
2021-09-03 20:22:05.043620: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
2021-09-03 20:22:08.828987: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
2021-09-03 20:22:12.362935: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
Horovod v0.22.1:

Available Frameworks:
    [X] TensorFlow
    [X] PyTorch
    [ ] MXNet

Available Controllers:
    [ ] MPI
    [X] Gloo

Available Tensor Operations:
    [ ] NCCL
    [ ] DDL
    [ ] CCL
    [ ] MPI
    [X] Gloo
```

It seems to be detecting the dual GPUs alright.

I am not familiar with parallel training, does this output look right?
I am just a little suspicious because I don't see the scaled learning rate in the output.

If so, maybe we could adding this Gloo method to the [parallel training](https://github.com/deepmodeling/deepmd-kit/blob/master/doc/train/parallel-training.md) doc page?

```
(base) [chazeon@exp-7-60 000.1]$ horovodrun -np 2 dp train params.yml
2021-09-03 12:24:35.436281: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
2021-09-03 12:24:43.160150: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
[1]<stderr>:2021-09-03 12:24:47.763641: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
[0]<stderr>:2021-09-03 12:24:47.775190: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
[0]<stderr>:WARNING:tensorflow:From /expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/lib/python3.9/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
[0]<stderr>:Instructions for updating:
[0]<stderr>:non-resource variables are not supported in the long term
[1]<stderr>:WARNING:tensorflow:From /expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/lib/python3.9/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
[1]<stderr>:Instructions for updating:
[1]<stderr>:non-resource variables are not supported in the long term
[1]<stderr>:WARNING:root:Environment variable KMP_BLOCKTIME is empty. Use the default value 0
[0]<stderr>:WARNING:root:Environment variable KMP_BLOCKTIME is empty. Use the default value 0
[1]<stderr>:WARNING:root:Environment variable KMP_AFFINITY is empty. Use the default value granularity=fine,verbose,compact,1,0
[0]<stderr>:WARNING:root:Environment variable KMP_AFFINITY is empty. Use the default value granularity=fine,verbose,compact,1,0
[0]<stderr>:/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/lib/python3.9/site-packages/deepmd/utils/compat.py:316: UserWarning: It seems that you are using a deepmd-kit input of version 1.x.x, which is deprecated. we have converted the input to >2.0.0 compatible, and output it to file input_v2_compat.json
[0]<stderr>:  warnings.warn(msg)
[1]<stderr>:/expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/lib/python3.9/site-packages/deepmd/utils/compat.py:316: UserWarning: It seems that you are using a deepmd-kit input of version 1.x.x, which is deprecated. we have converted the input to >2.0.0 compatible, and output it to file input_v2_compat.json
[1]<stderr>:  warnings.warn(msg)
[0]<stderr>:2021-09-03 12:25:01.266902: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
[0]<stderr>:To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
[0]<stderr>:2021-09-03 12:25:01.268524: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcuda.so.1
[1]<stderr>:2021-09-03 12:25:01.283516: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
[1]<stderr>:To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
[1]<stderr>:2021-09-03 12:25:01.287273: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcuda.so.1
[0]<stderr>:2021-09-03 12:25:01.328338: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
[0]<stderr>:2021-09-03 12:25:01.328354: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]
[0]<stderr>:2021-09-03 12:25:01.328406: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
[1]<stderr>:2021-09-03 12:25:01.329141: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
[1]<stderr>:2021-09-03 12:25:01.329195: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]
[1]<stderr>:2021-09-03 12:25:01.329251: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
[0]<stderr>:2021-09-03 12:25:01.335226: I tensorflow/core/platform/profile_utils/cpu_utils.cc:114] CPU Frequency: 2500000000 Hz
[1]<stderr>:2021-09-03 12:25:01.348792: I tensorflow/core/platform/profile_utils/cpu_utils.cc:114] CPU Frequency: 2500000000 Hz
[0]<stderr>:OMP: Info #155: KMP_AFFINITY: Initial OS proc set respected: 8
[0]<stderr>:OMP: Info #216: KMP_AFFINITY: decoding x2APIC ids.
[0]<stderr>:OMP: Info #157: KMP_AFFINITY: 1 available OS procs
[0]<stderr>:OMP: Info #158: KMP_AFFINITY: Uniform topology
[0]<stderr>:OMP: Info #287: KMP_AFFINITY: topology layer "LL cache" is equivalent to "socket".
[0]<stderr>:OMP: Info #287: KMP_AFFINITY: topology layer "L3 cache" is equivalent to "socket".
[0]<stderr>:OMP: Info #287: KMP_AFFINITY: topology layer "L2 cache" is equivalent to "core".
[0]<stderr>:OMP: Info #287: KMP_AFFINITY: topology layer "L1 cache" is equivalent to "core".
[0]<stderr>:OMP: Info #192: KMP_AFFINITY: 1 socket x 1 core/socket x 1 thread/core (1 total cores)
[0]<stderr>:OMP: Info #218: KMP_AFFINITY: OS proc to physical thread map:
[0]<stderr>:OMP: Info #172: KMP_AFFINITY: OS proc 8 maps to socket 0 core 2 thread 0
[0]<stderr>:OMP: Info #254: KMP_AFFINITY: pid 38680 tid 38786 thread 1 bound to OS proc set 8
[0]<stderr>:OMP: Info #254: KMP_AFFINITY: pid 38680 tid 38787 thread 2 bound to OS proc set 8
[1]<stderr>:OMP: Info #155: KMP_AFFINITY: Initial OS proc set respected: 8
[1]<stderr>:OMP: Info #216: KMP_AFFINITY: decoding x2APIC ids.
[1]<stderr>:OMP: Info #157: KMP_AFFINITY: 1 available OS procs
[1]<stderr>:OMP: Info #158: KMP_AFFINITY: Uniform topology
[1]<stderr>:OMP: Info #287: KMP_AFFINITY: topology layer "LL cache" is equivalent to "socket".
[1]<stderr>:OMP: Info #287: KMP_AFFINITY: topology layer "L3 cache" is equivalent to "socket".
[1]<stderr>:OMP: Info #287: KMP_AFFINITY: topology layer "L2 cache" is equivalent to "core".
[1]<stderr>:OMP: Info #287: KMP_AFFINITY: topology layer "L1 cache" is equivalent to "core".
[1]<stderr>:OMP: Info #192: KMP_AFFINITY: 1 socket x 1 core/socket x 1 thread/core (1 total cores)
[1]<stderr>:OMP: Info #218: KMP_AFFINITY: OS proc to physical thread map:
[1]<stderr>:OMP: Info #172: KMP_AFFINITY: OS proc 8 maps to socket 0 core 2 thread 0
[1]<stderr>:OMP: Info #254: KMP_AFFINITY: pid 38679 tid 38789 thread 1 bound to OS proc set 8
[1]<stderr>:OMP: Info #254: KMP_AFFINITY: pid 38679 tid 38790 thread 2 bound to OS proc set 8
[0]<stderr>:DEEPMD INFO    training data with min nbor dist: 1.0231442946289506
[0]<stderr>:DEEPMD INFO    training data with max nbor size: [64, 134, 64]
[1]<stderr>:DEEPMD INFO    training data with min nbor dist: 1.0231442946289506
[1]<stderr>:DEEPMD INFO    training data with max nbor size: [64, 134, 64]
[0]<stderr>:DEEPMD INFO     _____               _____   __  __  _____           _     _  _
[0]<stderr>:DEEPMD INFO    |  __ \             |  __ \ |  \/  ||  __ \         | |   (_)| |
[0]<stderr>:DEEPMD INFO    | |  | |  ___   ___ | |__) || \  / || |  | | ______ | | __ _ | |_
[0]<stderr>:DEEPMD INFO    | |  | | / _ \ / _ \|  ___/ | |\/| || |  | ||______|| |/ /| || __|
[0]<stderr>:DEEPMD INFO    | |__| ||  __/|  __/| |     | |  | || |__| |        |   < | || |_
[0]<stderr>:DEEPMD INFO    |_____/  \___| \___||_|     |_|  |_||_____/         |_|\_\|_| \__|
[0]<stderr>:DEEPMD INFO    Please read and cite:
[0]<stderr>:DEEPMD INFO    Wang, Zhang, Han and E, Comput.Phys.Comm. 228, 178-184 (2018)
[0]<stderr>:DEEPMD INFO    installed to:         /tmp/pip-req-build-m0vk2rxi/_skbuild/linux-x86_64-3.9/cmake-install
[0]<stderr>:DEEPMD INFO    source :              v2.0.0
[0]<stderr>:DEEPMD INFO    source brach:         HEAD
[0]<stderr>:DEEPMD INFO    source commit:        1a25414
[0]<stderr>:DEEPMD INFO    source commit at:     2021-08-28 08:15:38 +0800
[0]<stderr>:DEEPMD INFO    build float prec:     double
[0]<stderr>:DEEPMD INFO    build with tf inc:    /expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/lib/python3.9/site-packages/tensorflow/include
[0]<stderr>:DEEPMD INFO    build with tf lib:
[0]<stderr>:DEEPMD INFO    ---Summary of the training---------------------------------------
[0]<stderr>:DEEPMD INFO    distributed
[0]<stderr>:DEEPMD INFO    world size:           2
[0]<stderr>:DEEPMD INFO    my rank:              0
[0]<stderr>:DEEPMD INFO    node list:            ['exp-7-60']
[0]<stderr>:DEEPMD INFO    running on:           exp-7-60
[0]<stderr>:DEEPMD INFO    computing device:     gpu:0
[0]<stderr>:DEEPMD INFO    CUDA_VISIBLE_DEVICES: 0,1
[0]<stderr>:DEEPMD INFO    Count of visible GPU: 2
[0]<stderr>:DEEPMD INFO    num_intra_threads:    0
[0]<stderr>:DEEPMD INFO    num_inter_threads:    0
[0]<stderr>:DEEPMD INFO    -----------------------------------------------------------------
[0]<stderr>:2021-09-03 12:25:08.343280: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
[0]<stderr>:2021-09-03 12:25:08.343297: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]
[1]<stderr>:DEEPMD INFO     _____               _____   __  __  _____           _     _  _
[1]<stderr>:DEEPMD INFO    |  __ \             |  __ \ |  \/  ||  __ \         | |   (_)| |
[1]<stderr>:DEEPMD INFO    | |  | |  ___   ___ | |__) || \  / || |  | | ______ | | __ _ | |_
[1]<stderr>:DEEPMD INFO    | |  | | / _ \ / _ \|  ___/ | |\/| || |  | ||______|| |/ /| || __|
[1]<stderr>:DEEPMD INFO    | |__| ||  __/|  __/| |     | |  | || |__| |        |   < | || |_
[1]<stderr>:DEEPMD INFO    |_____/  \___| \___||_|     |_|  |_||_____/         |_|\_\|_| \__|
[1]<stderr>:DEEPMD INFO    Please read and cite:
[1]<stderr>:DEEPMD INFO    Wang, Zhang, Han and E, Comput.Phys.Comm. 228, 178-184 (2018)
[1]<stderr>:DEEPMD INFO    installed to:         /tmp/pip-req-build-m0vk2rxi/_skbuild/linux-x86_64-3.9/cmake-install
[1]<stderr>:DEEPMD INFO    source :              v2.0.0
[1]<stderr>:DEEPMD INFO    source brach:         HEAD
[1]<stderr>:DEEPMD INFO    source commit:        1a25414
[1]<stderr>:DEEPMD INFO    source commit at:     2021-08-28 08:15:38 +0800
[1]<stderr>:DEEPMD INFO    build float prec:     double
[1]<stderr>:DEEPMD INFO    build with tf inc:    /expanse/lustre/scratch/chazeon/temp_project/deepmd-kit-2.0.0-cuda10.1_gpu/lib/python3.9/site-packages/tensorflow/include
[1]<stderr>:DEEPMD INFO    build with tf lib:
[1]<stderr>:DEEPMD INFO    ---Summary of the training---------------------------------------
[1]<stderr>:DEEPMD INFO    distributed
[1]<stderr>:DEEPMD INFO    world size:           2
[1]<stderr>:DEEPMD INFO    my rank:              1
[1]<stderr>:DEEPMD INFO    node list:            ['exp-7-60']
[1]<stderr>:DEEPMD INFO    running on:           exp-7-60
[1]<stderr>:DEEPMD INFO    computing device:     gpu:1
[1]<stderr>:DEEPMD INFO    CUDA_VISIBLE_DEVICES: 0,1
[1]<stderr>:DEEPMD INFO    Count of visible GPU: 2
[1]<stderr>:DEEPMD INFO    num_intra_threads:    0
[1]<stderr>:DEEPMD INFO    num_inter_threads:    0
[1]<stderr>:DEEPMD INFO    -----------------------------------------------------------------
[1]<stderr>:2021-09-03 12:25:08.351655: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
[1]<stderr>:2021-09-03 12:25:08.351704: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]
[0]<stderr>:DEEPMD INFO    ---Summary of DataSystem: training     -----------------------------------------------
[0]<stderr>:DEEPMD INFO    found 24 system(s):
[0]<stderr>:DEEPMD INFO                                        system  natoms  bch_sz   n_bch   prob  pbc
[0]<stderr>:DEEPMD INFO                       ../data/init/004/V112.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/004/V114.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/004/V110.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/004/V116.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/003/V112.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/003/V110.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/003/V114.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/003/V116.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/001/V116.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/001/V110.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/001/V114.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/001/V112.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/002/V110.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/002/V116.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/002/V112.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                       ../data/init/002/V114.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                        ../data/init/005/V44.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                        ../data/init/005/V46.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO    ---Summary of DataSystem: training     -----------------------------------------------
[0]<stderr>:DEEPMD INFO                        ../data/init/005/V36.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                        ../data/init/005/V48.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                        ../data/init/005/V38.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO    found 24 system(s):
[0]<stderr>:DEEPMD INFO                        ../data/init/005/V42.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO                        ../data/init/005/V50.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                                        system  natoms  bch_sz   n_bch   prob  pbc
[0]<stderr>:DEEPMD INFO                        ../data/init/005/V40.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/004/V112.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO    --------------------------------------------------------------------------------------
[1]<stderr>:DEEPMD INFO                       ../data/init/004/V114.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/004/V110.0     128       8       5  0.042    T
[0]<stderr>:DEEPMD INFO    training without frame parameter
[1]<stderr>:DEEPMD INFO                       ../data/init/004/V116.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/003/V112.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/003/V110.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/003/V114.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/003/V116.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/001/V116.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/001/V110.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/001/V114.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/001/V112.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/002/V110.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/002/V116.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/002/V112.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                       ../data/init/002/V114.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                        ../data/init/005/V44.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                        ../data/init/005/V46.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                        ../data/init/005/V36.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                        ../data/init/005/V48.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                        ../data/init/005/V38.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                        ../data/init/005/V42.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                        ../data/init/005/V50.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO                        ../data/init/005/V40.0     128       8       5  0.042    T
[1]<stderr>:DEEPMD INFO    --------------------------------------------------------------------------------------
[1]<stderr>:DEEPMD INFO    training without frame parameter
[0]<stderr>:OMP: Info #254: KMP_AFFINITY: pid 38680 tid 38680 thread 0 bound to OS proc set 8
[1]<stderr>:OMP: Info #254: KMP_AFFINITY: pid 38679 tid 38679 thread 0 bound to OS proc set 8
[0]<stderr>:2021-09-03 12:25:36.835262: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
[0]<stderr>:2021-09-03 12:25:36.835328: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]
[1]<stderr>:2021-09-03 12:25:36.848644: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
[1]<stderr>:2021-09-03 12:25:36.848704: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]
[0]<stderr>:DEEPMD INFO    training data with min nbor dist: 1.0231442946289506
[0]<stderr>:DEEPMD INFO    training data with max nbor size: [64, 134, 64]
[0]<stderr>:DEEPMD INFO    built lr
[1]<stderr>:DEEPMD INFO    training data with min nbor dist: 1.0231442946289506
[1]<stderr>:DEEPMD INFO    training data with max nbor size: [64, 134, 64]
[1]<stderr>:DEEPMD INFO    built lr
[0]<stderr>:DEEPMD INFO    built network
[1]<stderr>:DEEPMD INFO    built network
[0]<stderr>:DEEPMD INFO    built training
[0]<stderr>:2021-09-03 12:25:53.598518: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties:
[0]<stderr>:pciBusID: 0000:3b:00.0 name: Tesla V100-SXM2-32GB computeCapability: 7.0
[0]<stderr>:coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 31.75GiB deviceMemoryBandwidth: 836.37GiB/s
[0]<stderr>:2021-09-03 12:25:53.598609: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
[0]<stderr>:2021-09-03 12:25:53.604333: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.10
[0]<stderr>:2021-09-03 12:25:53.604429: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublasLt.so.10
[0]<stderr>:2021-09-03 12:25:53.608642: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcufft.so.10
[0]<stderr>:2021-09-03 12:25:53.612354: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcurand.so.10
[0]<stderr>:2021-09-03 12:25:53.618215: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcusolver.so.10
[0]<stderr>:2021-09-03 12:25:53.620234: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcusparse.so.10
[0]<stderr>:2021-09-03 12:25:53.624802: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudnn.so.7
[0]<stderr>:2021-09-03 12:25:53.629973: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1871] Adding visible gpu devices: 0
[0]<stderr>:2021-09-03 12:25:53.630038: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
[1]<stderr>:DEEPMD INFO    built training
[1]<stderr>:2021-09-03 12:25:53.777451: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties:
[1]<stderr>:pciBusID: 0000:86:00.0 name: Tesla V100-SXM2-32GB computeCapability: 7.0
[1]<stderr>:coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 31.75GiB deviceMemoryBandwidth: 836.37GiB/s
[1]<stderr>:2021-09-03 12:25:53.777573: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
[1]<stderr>:2021-09-03 12:25:53.781096: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.10
[1]<stderr>:2021-09-03 12:25:53.781559: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublasLt.so.10
[1]<stderr>:2021-09-03 12:25:53.786484: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcufft.so.10
[1]<stderr>:2021-09-03 12:25:53.789169: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcurand.so.10
[1]<stderr>:2021-09-03 12:25:53.806169: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcusolver.so.10
[1]<stderr>:2021-09-03 12:25:53.807383: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcusparse.so.10
[1]<stderr>:2021-09-03 12:25:53.811167: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudnn.so.7
[1]<stderr>:2021-09-03 12:25:53.814225: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1871] Adding visible gpu devices: 1
[1]<stderr>:2021-09-03 12:25:53.814307: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.10.1
[0]<stderr>:2021-09-03 12:25:54.631211: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
[0]<stderr>:2021-09-03 12:25:54.632627: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      0
[0]<stderr>:2021-09-03 12:25:54.632705: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 0:   N
[0]<stderr>:2021-09-03 12:25:54.640958: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1418] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 31006 MB memory) -> physical GPU (device: 0, name: Tesla V100-SXM2-32GB, pci bus id: 0000:3b:00.0, compute capability: 7.0)
[1]<stderr>:2021-09-03 12:25:54.807253: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
[1]<stderr>:2021-09-03 12:25:54.807574: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      1
[1]<stderr>:2021-09-03 12:25:54.807640: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 1:   N
[1]<stderr>:2021-09-03 12:25:54.816411: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1418] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 31006 MB memory) -> physical GPU (device: 1, name: Tesla V100-SXM2-32GB, pci bus id: 0000:86:00.0, compute capability: 7.0)
[0]<stderr>:DEEPMD INFO    initialize model from scratch
[1]<stderr>:DEEPMD INFO    receive global variables from task#0
[0]<stderr>:DEEPMD INFO    broadcast global variables to other tasks
[1]<stderr>:DEEPMD INFO    start training at lr 5.00e-04 (== 5.00e-04), decay_step 5000, decay_rate 0.763002, final lr will be 1.00e-08
[0]<stderr>:DEEPMD INFO    start training at lr 5.00e-04 (== 5.00e-04), decay_step 5000, decay_rate 0.763002, final lr will be 1.00e-08
[0]<stderr>:2021-09-03 12:26:01.051921: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.10
[1]<stderr>:2021-09-03 12:26:05.779826: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.10
[1]<stderr>:DEEPMD INFO    batch     100 training time 18.69 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     100 training time 18.54 s, testing time 0.04 s
[1]<stderr>:DEEPMD INFO    batch     200 training time 11.48 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     200 training time 11.45 s, testing time 0.04 s
[1]<stderr>:DEEPMD INFO    batch     300 training time 12.61 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     300 training time 12.81 s, testing time 0.06 s
[1]<stderr>:DEEPMD INFO    batch     400 training time 11.69 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     400 training time 11.44 s, testing time 0.04 s
[1]<stderr>:DEEPMD INFO    batch     500 training time 12.62 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     500 training time 12.88 s, testing time 0.04 s
[1]<stderr>:DEEPMD INFO    batch     600 training time 11.70 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     600 training time 11.55 s, testing time 0.04 s
[1]<stderr>:DEEPMD INFO    batch     700 training time 11.48 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     700 training time 11.41 s, testing time 0.04 s
[1]<stderr>:DEEPMD INFO    batch     800 training time 12.56 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     800 training time 12.80 s, testing time 0.06 s
[1]<stderr>:DEEPMD INFO    batch     900 training time 11.72 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch     900 training time 11.42 s, testing time 0.04 s
[1]<stderr>:DEEPMD INFO    batch    1000 training time 11.49 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch    1000 training time 11.50 s, testing time 0.04 s
[0]<stderr>:DEEPMD INFO    saved checkpoint model.ckpt
[1]<stderr>:DEEPMD INFO    batch    1100 training time 11.65 s, testing time 0.00 s
[0]<stderr>:DEEPMD INFO    batch    1100 training time 12.84 s, testing time 0.06 s
[1]<stderr>:DEEPMD INFO    batch    1200 training time 12.69 s, testing time 0.00 s
```

**Comment by njzjz** [comment]
Hi @chazeon, have you tested the horovod/horovod#3148? If this also works for you, I plan to build horovod with this patch in next deepmd-kit release package.

**Comment by shishaochen** [comment]
@njzjz With your fix, the water example can train in parallel with Horovod over MPICH.
At my dev host, occupied GPU cards are running as expected.

cc @chazeon

**Comment by chazeon** [comment]
@njzjz I am having other trouble with MPICH's MPI I am trying to figure out what's wrong...
I am working on a HPC's that has very slow IO, sorry for the delay.

**Comment by chazeon** [comment]
In summary -- the fix on horovod goes in the right direction in general, but there are actually more complications...

Because OpenMPI relies on MCA parameters set to environments to determine most of the runtime settings, whereas MPICH relies on command-line arguments, builds with MPICH require more careful treatments.

In my case, with MPICH, I have to manually choose a launcher, the mpirun shipped in the v2.0.0 release support slurm (`srun`), ssh, fork, etc., see "Launchers available" entry below:
```
(base) [chazeon@exp-7-59 ~]$ mpirun --version
HYDRA build details:
    Version:                                 3.3.2
    Release Date:                            Tue Nov 12 21:23:16 CST 2019
    ...
    Process Manager:                         pmi
    Launchers available:                     ssh rsh fork slurm ll lsf sge manual persist
    Topology libraries available:            hwloc
    Resource management kernels available:   user slurm ll lsf sge pbs cobalt
    Checkpointing libraries available:
    Demux engines available:                 poll select
```
 I need to use `ssh` or `fork`, but `mpirun` default to `srun`, so I will need to set the [`-launcher` parameter](https://wiki.mpich.org/mpich/index.php/Using_the_Hydra_Process_Manager#Resource_Managers_and_Launchers) manually.

So I will have to manually change `horovod/runner/mpi_run.py` Line 200 from
```python
basic_args = '-l' if impi else '--allow-run-as-root --tag-output'
```
to
```python
basic_args = '-l -launcher ssh' if impi else '--allow-run-as-root --tag-output'
```
to mimic how I manually invoke `mpirun`.

In OpenMPI this was achieved by setting [`pls_rsh_agent`](https://www.open-mpi.org/faq/?category=rsh#rsh-not-ssh) and an example can be seen used in [`horovod/spark/mpi_run.py`](https://github.com/horovod/horovod/blob/8186cad7c7c8ca0d03334be823b971d83c386a93/horovod/spark/mpi_run.py).

Nevertheless, with this manual alternation, I am able to run the two tests and the training.

In my overall experience, working with Gloo is actually much more straightforward and I don't see there is a significant performance difference overall, although the `--mpi-log` is useless in that case.

**Comment by chazeon** [comment]
In [`deepmd/loggers/loggers.py`](https://github.com/deepmodeling/deepmd-kit/blob/1a25414dcd18b12316250baa330b690f3f5510fe/deepmd/loggers/loggers.py), MPI-based `rank` check could be changed to use Horovodrun's API to support.

**Comment by chazeon** [comment]
I realize I can still [manually invoke `mpirun` command manually](https://github.com/horovod/horovod/blob/master/docs/mpi.rst) and horovod can still be utilized
```bash
mpirun -np 2 -launcher rsh -hosts=localhost dp train params.yml
```
Then I am satisfied with the current solution.

**Comment by njzjz** [comment]
horovod is packaged in the new version of installer.
